modules/json/src/smithy4s/http/json/SchematicJCodec.scala

```diff
@@ -464,76 +466,163 @@ private[smithy4s] class SchematicJCodec(constraints: Constraints, maxArity: Int)
+      }
 
-        val altCache = new PolyFunction[Alt[JCodecMake, Z, *], JCodec] {
-          def apply[A](fa: Alt[JCodecMake, Z, A]): JCodec[A] = fa.instance.get
-        }.unsafeCache((first +: rest).map(alt => Existential.wrap(alt)))
+      val altCache = new PolyFunction[Alt[JCodecMake, Z, *], JCodec] {
+        def apply[A](fa: Alt[JCodecMake, Z, A]): JCodec[A] = fa.instance.get
+      }.unsafeCache((first +: rest).map(alt => Existential.wrap(alt)))
 
-        def encodeValue(z: Z, out: JsonWriter): Unit = {
-          def writeValue[A](awv: Alt.WithValue[JCodecMake, Z, A]): Unit =
-            altCache(awv.alt).encodeValue(awv.value, out)
+      def encodeValue(z: Z, out: JsonWriter): Unit = {
+        def writeValue[A](awv: Alt.WithValue[JCodecMake, Z, A]): Unit =
+          altCache(awv.alt).encodeValue(awv.value, out)
 
-          val awv = total(z)
-          out.writeObjectStart()
-          out.writeKey(jsonLabel(awv.alt))
-          writeValue(awv)
-          out.writeObjectEnd()
+        val awv = total(z)
+        out.writeObjectStart()
+        out.writeKey(jsonLabel(awv.alt))
+        writeValue(awv)
+        out.writeObjectEnd()
+      }
+
+      def decodeKey(in: JsonReader): Z =
+        in.decodeError("Cannot use coproducts as keys")
+
+      def encodeKey(x: Z, out: JsonWriter): Unit =
+        out.encodeError("Cannot use coproducts as keys")
+    }
+
+  private def discriminatedUnion[Z](
+      first: Alt[JCodecMake, Z, _],
+      rest: Vector[Alt[JCodecMake, Z, _]],
+      discriminated: Discriminated
+  )(total: Z => Alt.WithValue[JCodecMake, Z, _]): JCodec[Z] =
+    new JCodec[Z] {
+
+      override lazy val expecting: String = "discriminated-union"
+
+      override def canBeKey: Boolean = false
+
+      def jsonLabel[A](alt: Alt[JCodecMake, Z, A]): String =
+        alt.instance.hints.get(JsonName).map(_.value).getOrElse(alt.label)
+
+      val handlerMap: Map[String, (Cursor, JsonReader) => Z] = {
+        val res = MMap.empty[String, (Cursor, JsonReader) => Z]
+        def handler[A](alt: Alt[JCodecMake, Z, A]) = {
+          val codec = alt.instance.get
+          (cursor: Cursor, reader: JsonReader) =>
+            alt.inject(cursor.decode(codec, reader))
         }
 
-        def decodeKey(in: JsonReader): Z =
-          in.decodeError("Cannot use coproducts as keys")
+        res += (jsonLabel(first) -> handler(first))
+        rest.foreach(alt => res += (jsonLabel(alt) -> handler(alt)))
+        res.toMap
+      }
 
-        def encodeKey(x: Z, out: JsonWriter): Unit =
-          out.encodeError("Cannot use coproducts as keys")
+      def decodeValue(cursor: Cursor, in: JsonReader): Z = {
+        if (in.isNextToken('{')) {
+          if (in.isNextToken('}'))
+            in.decodeError("Expected at least a one key/value pair")
+          else {
+            in.rollbackToken()
+            in.setMark()
+            val key = if (in.skipToKey(discriminated.propertyName)) {
```

Comment:
oh really cool ! I thought we were gonna have to do that by hand, happy it's not the case

Reply:
Yeah I was super happy to learn of `setMark` + `skipToKey` + `rollbackToMark`.. made the backtracking a lot easier

Reply:
Yeah I was aware of `setMark` + `rollbackToMark`. Kudos to @plokhotnyuk, I **really** appreciate jsoniter's API.